sympy/diffgeom/diffgeom.py

```diff
@@ -1084,6 +1086,388 @@ def __call__(self, field):
+    >>> TP=TensorProduct
+    >>> metric = f(r)**2*TP(dt, dt) - f(r)**(-2)*TP(dr, dr) - r**2*TP(dtheta, dtheta) - r**2*sin(theta)**2*TP(dphi, dphi)
+    >>> rm = TensorArray(components=metric_to_Riemann_components(metric), variance=[-1,1,1,1],coordinate_system=cs)  #Note the variance!
+
+    """
+    def _contravariant_slots(self,T):
+        slots=[]
+        for k in range(self.order):
+            l = [None for i in range(self.order-1)]
+            l.insert(k,self.base_vectors[0])
+            try:
+                T.rcall(*l)
+            except:
+                slots.append(k)
+        return slots
+
+    def __init__(self, components, variance, coordinate_system):
+        self.CoordSystem = coordinate_system
+        self.coords = self.CoordSystem.coord_functions()
+        self.base_oneforms = self.CoordSystem.base_oneforms()
+        self.base_vectors = self.CoordSystem.base_vectors()
+        self.n = len(self.coords)
+        self.order = len(variance)
+        self.shape = [self.n for _ in range(self.order)]
+        self.covariant_slots=[k for k in range(self.order) if variance[k]==1]
+        self.covariant_order = len(self.covariant_slots)
+        self.contravariant_slots=[k for k in range(self.order) if variance[k]==-1]
+        self.contravariant_order = len(self.contravariant_slots)
+        #Workaround for a bug in the NDimArray class:
+        if self.order == 0:
+            self.tensor = copy.copy(components)
+            self.indices = [0]
+        else:
+            self.indices = list(itertools.product(range(self.n),repeat=self.order))
+            self.tensor = MutableSparseNDimArray.zeros(*self.shape)
+            components = MutableSparseNDimArray(components)
+            for index in self.indices:
+                if components[index]!=0:
+                    self.tensor[index] = components[index]
+                else:
+                    self.indices.remove(index)
+
+    def to_tensor(self):
+        if self.shape == []:
+            #Sympy's tensor classes are broken
+            return self.tensor._sparse_array[0]
+        def dualbasishelper(p,k):
+            if self.contravariant_slots.count(k) == 0:
+                return self.base_oneforms[p[k]]
+            else:
+                return self.base_vectors[p[k]]
+        Tbasesdual = {
+            I : TensorProduct(*[dualbasishelper(I,k) for k in range(self.order)])
+            for I in self.indices
+        }
+        return sum([Tbasesdual[I] * self.tensor[I] for I in self.indices])
+
+    def components(self):
+        """
+        Nonzero components
+        """
+        return [self.tensor[index] for index in self.indices]
+
+    def variance(self,index):
+        if self.contravariant_slots.count(index) > 0:
+            return 'contravariant'
+        elif self.covariant_slots.count(index) > 0:
+            return 'covariant'
+        else:
+            raise IndexError
+
+    def _TensorProduct(self,other): #destructive tensor product
```

Comment:
there is the `tensorproduct` function is `sympy.tensor.array`. It can be combined with `tensorcontraction`.